### PR TITLE
add an unsynchronised LRU for the projection cache

### PR DESCRIPTION
**Pull request by kinnou02**

### CODE REVIEW

File: asgard/projector.cpp [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+}
+
+Projector::Projector(size_t cache_size): cache_size(cache_size) {
+    if (cache_size < 1) {
+        throw std::invalid_argument("max (size of cache) must be strictly positive");
+    }
+}
+
+std::unordered_map<std::string, valhalla::baldr::PathLocation>
+Projector::operator()(const std::vector<std::string>& places,
```
Comment (antoine-de) [inline asgard/projector.cpp]: strange indent

File: asgard/projector.h [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+#include <boost/multi_index/member.hpp>
+
+namespace asgard{
+
+
+class wrong_coordinate: public std::runtime_error{using runtime_error::runtime_error;};
+
+class Projector {
+private:
+    typedef typename std::string key_type;
```
Comment (antoine-de) [inline asgard/projector.h]: the typename is mandatory here ?

File: asgard/asgard.cpp [hunk 3]
```diff
@@ -101,50 +100,33 @@ void worker(zmq::context_t& context, valhalla::baldr::GraphReader& graph){
...
-                loc.name_ = e.place();
-                targets.push_back(loc);
-            }else{
-                path_location_targets.push_back(it->second);
-            }
+            targets.push_back(e.place());
         }
 
-        std::vector<valhalla::baldr::Location> locations(sources);
+        std::vector<std::string> locations(sources);
```
Comment (antoine-de) [inline asgard/asgard.cpp]: isn't it possible for the projector to work with iterators ?

**Comment by antoine-de** [comment]
I think it's better to only have one way to handle coord

so `lon:lat`

here is the code (to move in utils?) handling that in kraken
https://github.com/CanalTP/navitia/blob/dev/source/type/type.cpp#L1075-L1082

File: asgard/asgard.cpp [hunk 4]
```diff
@@ -101,50 +100,33 @@ void worker(zmq::context_t& context, valhalla::baldr::GraphReader& graph){
...
         locations.insert(end(locations), begin(targets), end(targets));
-        LOG_INFO((boost::format("cache ratio %s/%s") % (path_location_sources.size() + path_location_targets.size()) % (path_location_sources.size() + path_location_targets.size() + locations.size())).str());
         LOG_INFO("projecting");
         auto costing = mode_costing[static_cast<int>(mode_map[pb_req.sn_routing_matrix().mode()])];
```
Comment (TeXitoi) [inline asgard/asgard.cpp]: 💩

File: asgard/projector.h [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+            throw std::invalid_argument("max (size of cache) must be strictly positive");
+        }
+    }
+
+    template <typename T>
+    std::unordered_map<std::string,
+                      valhalla::baldr::PathLocation> operator()(const T places_begin, const T places_end,
+                                                                valhalla::baldr::GraphReader& graph,
+                                                                valhalla::sif::cost_ptr_t costing) const{
+    std::unordered_map<std::string, valhalla::baldr::PathLocation> results;
```
Comment (antoine-de) [inline asgard/projector.h]: don't you need another indent level here ?

File: asgard/projector.h [hunk 6]
```diff
@@ -10,27 +10,19 @@
 #include <boost/multi_index/ordered_index.hpp>
 #include <boost/multi_index/hashed_index.hpp>
 #include <boost/multi_index/member.hpp>
+#include "utils/coord_parser.h"
 
 namespace asgard{
 
-
-class wrong_coordinate: public std::runtime_error{using runtime_error::runtime_error;};
+using wrong_coordinate = wrong_coordinate;
```
Comment (antoine-de) [inline asgard/projector.h]: is it useful ?

Comment (kinnou02) [inline asgard/projector.h]: This way it's namespaced in asgard

**Pull request merged**